### PR TITLE
[TRA-62] Add market type to the CreatePerpetual API

### PR DESCRIPTION
**Pull request by shrenujb**
### Changelist
Add a Perpetual Market Type to denote if a market is cross or isolated
Add this market type to the CreatePerpetual API
Make all resulting API changes

### Test Plan
[Describe how this PR was tested (if applicable)]

### Author/Reviewer Checklist
- [x] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [x] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [x] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [x] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

### CODE REVIEW

File: protocol/x/perpetuals/types/types.go [hunk 1]
```diff
@@ -18,6 +18,25 @@ type GetPricePremiumParams struct {
 	MaxAbsPremiumVotePpm        *big.Int
 }
 
+//// PerpetualMarketType enum
```
Comment (vincentwschau) [inline protocol/x/perpetuals/types/types.go:21]: Why is this commented out?

File: protocol/x/perpetuals/keeper/perpetual_test.go [hunk 2]
```diff
@@ -142,6 +143,7 @@ func TestCreatePerpetual_Failure(t *testing.T) {
 		atomicResolution  int32
 		defaultFundingPpm int32
 		liquidityTier     uint32
+		marketType        types.PerpetualMarketType
```
Comment (vincentwschau) [inline protocol/x/perpetuals/keeper/perpetual_test.go:146]: Add a test for an invalid market type leading to the `CreatePerpetual` call failing?

File: protocol/testutil/keeper/perpetuals.go [hunk 3]
```diff
@@ -236,6 +239,7 @@ func CreateNPerpetuals(
 			int32(i),             // AtomicResolution
 			defaultFundingPpm,    // DefaultFundingPpm
 			allLiquidityTiers[i%len(allLiquidityTiers)].Id, // LiquidityTier
+			types.PerpetualMarketType_PERPETUAL_MARKET_TYPE_CROSS, // MarketType
```
Comment (vincentwschau) [inline protocol/testutil/keeper/perpetuals.go:242]: nit: Alternate between cross and isolated / create at least 1 isolated perpetual.

File: protocol/testutil/keeper/perpetuals.go [hunk 4]
```diff
@@ -285,6 +289,7 @@ func CreateTestPricesAndPerpetualMarkets(
 			perp.Params.AtomicResolution,
 			perp.Params.DefaultFundingPpm,
 			perp.Params.LiquidityTier,
+			types.PerpetualMarketType_PERPETUAL_MARKET_TYPE_CROSS,
```
Comment (vincentwschau) [inline protocol/testutil/keeper/perpetuals.go:292]: Shouldn't this be pulled from the `perpetuals` passed in rather than being hard-coded?

File: protocol/testutil/keeper/perpetuals.go (hunk 3 above)
Comment (shrenujb) [reply protocol/testutil/keeper/perpetuals.go:242]: not sure I understand the comment. Doesn't this test need all markets to be cross?

File: protocol/x/clob/keeper/mev_test.go [hunk 5]
```diff
@@ -880,6 +880,7 @@ func TestRecordMevMetrics(t *testing.T) {
 					p.Params.AtomicResolution,
 					p.Params.DefaultFundingPpm,
 					p.Params.LiquidityTier,
+					p.Params.MarketType,
 				)
 				require.NoError(t, err)
 			}
```
[the remark this answers is not in the document]
Comment (shrenujb) [reply protocol/x/clob/keeper/mev_test.go:886]: Yes, I would like assistance in creating additional test cases

File: protocol/x/perpetuals/keeper/perpetual.go [hunk 6]
```diff
@@ -40,6 +40,7 @@ func (k Keeper) CreatePerpetual(
 	atomicResolution int32,
 	defaultFundingPpm int32,
 	liquidityTier uint32,
+	marketType types.PerpetualMarketType,
 ) (types.Perpetual, error) {
 	// Check if perpetual exists.
 	if k.HasPerpetual(ctx, id) {
```
[the remark this answers is not in the document]
Comment (shrenujb) [reply protocol/x/perpetuals/keeper/perpetual.go:46]: Yes

File: protocol/testutil/keeper/perpetuals.go (hunk 3 above)
Comment (vincentwschau) [reply protocol/testutil/keeper/perpetuals.go:242]: I don't think this is a test, it's a test util function? If you change this to be say CROSS for half of the N perps and ISOLATED for the other half what tests break?

File: protocol/x/perpetuals/keeper/perpetual_test.go (hunk 2 above)
Comment (shrenujb) [reply protocol/x/perpetuals/keeper/perpetual_test.go:146]: done

File: protocol/testutil/constants/genesis.go [hunk 7]
```diff
@@ -901,7 +901,8 @@ const GenesisState = `{
             "id": 0,
             "liquidity_tier": 0,
             "market_id": 0,
-            "ticker": "BTC-USD"
+            "ticker": "BTC-USD",
+		"market_type": 1
```
Comment (vincentwschau) [inline protocol/testutil/constants/genesis.go:905]: nit: Remove indent. Same for below.

File: protocol/x/perpetuals/module_test.go [hunk 8]
```diff
@@ -271,7 +272,8 @@ func TestAppModule_InitExportGenesis(t *testing.T) {
 			  "params": {
 				 "ticker":"EXAM-USD",
 				 "market_id":0,
-				 "liquidity_tier":0
+				 "liquidity_tier":0,
+                 "market_type":"PERPETUAL_MARKET_TYPE_CROSS"
```
Comment (vincentwschau) [inline protocol/x/perpetuals/module_test.go:276]: nit: Match the indent. Same for below.

File: protocol/x/perpetuals/client/cli/query_perpetual_test.go [hunk 9]
```diff
@@ -66,6 +66,7 @@ func networkWithLiquidityTierAndPerpetualObjects(
 				Id:            uint32(i),
 				Ticker:        fmt.Sprintf("test_query_ticker_%d", i),
 				LiquidityTier: uint32(i % m),
+				MarketType:    types.PerpetualMarketType_PERPETUAL_MARKET_TYPE_CROSS,
```
Comment (vincentwschau) [inline protocol/x/perpetuals/client/cli/query_perpetual_test.go:69]: nit: Similar to the create n perpetuals, have this change between cross and isolated. Makes it so the test below will be testing querying isolated perpetuals.

File: protocol/x/perpetuals/keeper/msg_server_update_perpetual_params_test.go [hunk 10]
```diff
@@ -192,7 +193,19 @@ func TestUpdatePerpetualParams(t *testing.T) {
 				// Verify updated perpetual params in state.
 				updatedPerpetualInState, err := pc.PerpetualsKeeper.GetPerpetual(pc.Ctx, tc.msg.PerpetualParams.Id)
 				require.NoError(t, err)
-				require.Equal(t, tc.msg.PerpetualParams, updatedPerpetualInState.Params)
+				require.Equal(t, tc.msg.PerpetualParams.Ticker, updatedPerpetualInState.Params.Ticker)
```
Comment (vincentwschau) [inline protocol/x/perpetuals/keeper/msg_server_update_perpetual_params_test.go:196]: nit: Get initial perpetual and verify that the market type is the same as the initial.

**Review by vincentwschau — DISMISSED** [review]

**Comment by vincentwschau** [comment]
You'll have to update the genesis json that's generated as well to include the market type [here](https://github.com/dydxprotocol/v4-chain/blob/main/protocol/testing/genesis.sh#L126) I think.

File: protocol/scripts/genesis/copy_sample_pregenesis.sh [hunk 11]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
Comment (vincentwschau) [inline protocol/scripts/genesis/copy_sample_pregenesis.sh:1]: Is this file intended to be included in the PR?
